vpn-network/vpn-network-impl/src/main/java/com/duckduckgo/vpn/network/impl/RealVpnNetwork.kt

```diff
@@ -0,0 +1,249 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.vpn.network.impl
+
+import android.content.Context
+import android.net.ConnectivityManager
+import android.net.VpnService
+import android.os.Build
+import android.os.Process
+import android.util.Log
+import androidx.annotation.RequiresApi
+import com.duckduckgo.appbuildconfig.api.AppBuildConfig
+import com.duckduckgo.di.scopes.VpnScope
+import com.duckduckgo.library.loader.LibraryLoader
+import com.duckduckgo.vpn.network.api.*
+import com.duckduckgo.vpn.network.impl.models.Allowed
+import com.squareup.anvil.annotations.ContributesBinding
+import dagger.SingleInstanceIn
+import com.duckduckgo.vpn.network.impl.models.Packet
+import com.duckduckgo.vpn.network.impl.models.ResourceRecord
+import com.duckduckgo.vpn.network.impl.models.Usage
+import timber.log.Timber
+import java.net.InetSocketAddress
+import java.util.concurrent.atomic.AtomicReference
+import javax.inject.Inject
+import kotlin.system.exitProcess
+
+@ContributesBinding(VpnScope::class)
+@SingleInstanceIn(VpnScope::class)
+class RealVpnNetwork @Inject constructor(
+    private val appBuildConfig: AppBuildConfig,
+    private val context: Context,
+) : VpnNetwork {
+
+    private external fun jni_init(sdk: Int): Long
+
+    private external fun jni_start(context: Long, loglevel: Int)
+
+    private external fun jni_run(context: Long, tun: Int, fwd53: Boolean, rcode: Int)
+
+    private external fun jni_stop(context: Long)
+
+    private external fun jni_clear(context: Long)
+
+    private external fun jni_get_mtu(): Int
+
+    private external fun jni_get_stats(context: Long): IntArray?
+
+    private external fun jni_pcap(name: String, record_size: Int, file_size: Int)
+
+    private external fun jni_socks5(addr: String, port: Int, username: String, password: String)
+
+    private external fun jni_done(context: Long)
+
+    private val uidPackageIdMap = mutableMapOf<Int, String>()
+
+    private var callback = AtomicReference<VpnNetworkCallback?>(null)
+
+    override fun create(): Long {
+        return jni_init(appBuildConfig.sdkInt)
+    }
+
+    override fun start(contextId: Long, logLevel: VpnNetworkLog) {
+        jni_socks5("", 0, "", "")
+        jni_start(contextId, logLevel.toAndroidLevel())
+    }
+
+    override fun run(contextId: Long, tunfd: Int) {
+        jni_run(contextId, tunfd, true, 3)
+    }
+
+    override fun stop(contextId: Long) {
+        jni_stop(contextId)
+    }
+
+    override fun destroy(contextId: Long) {
+//        jni_clear(contextId)
```

Comment:
`jni_clear` is called internally when calling `jni_done`. Leaving this commented code here to come back to it as I need to clean up the native networking layer